### ISSUE TITLE
Commit to redb based on memory instead of blocks processed

### ISSUE THREAD

**Issue by andrewtoth** [casey/ord#1630]
We currently commit to redb every 5000 blocks, after which our caches are cleared. For periods of low activity this uses little memory, and for periods of high activity it uses a lot of memory (4GB+). This is not ideal for either case. We want to keep our cache filled as long as possible but not use too much memory. This can be tricky with `HashMap` and `Vec` since they reallocate on their own and their capacity can be higher, but we can track our current entries' sizes and flush on a conservative amount to keep our memory usage to a certain level. Before flushing we can `shrink_to_fit` the caches as well before draining them.
Also, how much memory should we try and stay below by default? Should we allow users to set a higher/lower limit, like `dbcache` in `bitcoind`?

**Comment by raphjaph** [comment]
Yes, I like the idea of thinking in terms of cache and memory instead of blocks. For the default memory size we should just test to find a sweet spot. Maybe building on a raspberry pi could even work with this method.

Not using blocks as the flush points will probably make deterministic index builds(#1544) impossible but I think that is lower priority.

**Comment by veryordinally** [comment]
This would clearly make sense - when I wrote that code, the 5000 was just a random easy first guess at what might make sense. I benchmarked different numbers (1000, 10000), and 5000 seemed to be a good tradeoff on my machine.

Overall tradeoff is between indexing speed and how much indexing progress you lose in case of interruption. But if you have more memory and are willing to risk losing more progress if the indexing process gets interrupted, introducing a cache size constraint instead of a fixed # block trigger will make sense

**Comment by jorge-j1m** [comment]
I have the same issue, I think committing takes too much of the indexing time, I'm running the process in a fairly capable machine and I'd love to see more options to configure the ord process itself. Ideally we would match `bitcoind`'s `dbcache` size but I couldn't find any RPC to get that info, and reading from the `bitcoin.conf` is not a possibility in most cloud environments. Then here's my proposal, make the index commits dynamic, either by an ENV, a flag, or in the config file, something like `--index-memory` would make sense. If `dbcache` is not possible to get, then make the memory top default 2GB (we could test to find a more optimal, data based value, this is just what I think it makes sense). I can't make a PR but I can help test whatever you come up with, just reference this issue or tag me. Thanks @andrewtoth for the link to this one!

**Comment by raphjaph** [comment]
@partialord you're now pretty familiar with our indexing logic. Do you think this is an avenue worth exploring?

**Comment by partialord** [comment]
Yes, I think it's very likely worth doing something along these lines.  I'm currently traveling, but I'm planning to look at it when I get back on the 9th!
